src/rgw/rgw_lua_background.h

```diff
@@ -11,26 +13,142 @@ namespace rgw::lua {
 constexpr const int INIT_EXECUTE_INTERVAL = 5;
 
 //Writeable meta table named RGW with mutex protection
-using BackgroundMap = std::unordered_map<std::string, std::string>;
-struct RGWTable : StringMapMetaTable<BackgroundMap,
-  StringMapWriteableNewIndex<BackgroundMap>> {
-    static std::string TableName() {return "RGW";}
-    static std::string Name() {return TableName() + "Meta";}
-    static int IndexClosure(lua_State* L) {
-      auto& mtx = *reinterpret_cast<std::mutex*>(lua_touserdata(L, lua_upvalueindex(2)));
-      std::lock_guard l(mtx);
-      return StringMapMetaTable::IndexClosure(L);
+using BackgroundMapValue = std::variant<std::string, int64_t, double, bool>;
+using BackgroundMap  = std::unordered_map<std::string, BackgroundMapValue>;
+
+inline void pushvalue(lua_State* L, const std::string& value) {
+  pushstring(L, value);
+}
+
+inline void pushvalue(lua_State* L, int64_t value) {
+  lua_pushinteger(L, value);
+}
+
+inline void pushvalue(lua_State* L, double value) {
+  lua_pushnumber(L, value);
+}
+
+inline void pushvalue(lua_State* L, bool value) {
+  lua_pushboolean(L, value);
+}
+
+struct RGWTable : EmptyMetaTable {
+  static std::string TableName() {return "RGW";}
+  static std::string Name() {return TableName() + "Meta";}
+  
+  static int IndexClosure(lua_State* L) {
+    auto& mtx = *reinterpret_cast<std::mutex*>(lua_touserdata(L, lua_upvalueindex(2)));
+    const auto map = reinterpret_cast<BackgroundMap*>(lua_touserdata(L, lua_upvalueindex(1)));
+    const char* index = luaL_checkstring(L, 2);
+
+    std::lock_guard l(mtx);
+
+    const auto it = map->find(std::string(index));
+    if (it == map->end()) {
+      lua_pushnil(L);
+    } else {
+      std::visit([L](auto&& value) { pushvalue(L, value); }, it->second);
+    }
+    return ONE_RETURNVAL;
+  }
+
+  static int LenClosure(lua_State* L) {
+    auto& mtx = *reinterpret_cast<std::mutex*>(lua_touserdata(L, lua_upvalueindex(2)));
+    const auto map = reinterpret_cast<BackgroundMap*>(lua_touserdata(L, lua_upvalueindex(1)));
+
+    std::lock_guard l(mtx);
+
+    lua_pushinteger(L, map->size());
+
+    return ONE_RETURNVAL;
+  }
+
+  static int NewIndexClosure(lua_State* L) {
+    auto& mtx = *reinterpret_cast<std::mutex*>(lua_touserdata(L, lua_upvalueindex(2)));
+    const auto map = reinterpret_cast<BackgroundMap*>(lua_touserdata(L, lua_upvalueindex(1)));
+    const auto index = luaL_checkstring(L, 2);
+
+    std::unique_lock l(mtx);
+
+    size_t len;
+    BackgroundMapValue value;
+    const int value_type = lua_type(L, 3);
```

Comment:
and 3?

Reply:
these are positions in the lua stack. don;t see a good way to avoid these "magic numbers" as they are specific to each function